examples/android/clientcache/app/src/test/java/io/grpc/clientcacheexample/SafeMethodCachingInterceptorTest.java

```diff
@@ -0,0 +1,533 @@
+
+    assertEquals(reply1, reply2);
+  }
+
+  @Test
+  public void responseMaxAge_caseInsensitive() throws Exception {
+    cacheControlDirectives.add("MaX-aGe=0");
+
+    HelloReply reply1 =
+        ClientCalls.blockingUnaryCall(
+            channelToUse, safeGreeterSayHelloMethod, CallOptions.DEFAULT, message);
+    HelloReply reply2 =
+        ClientCalls.blockingUnaryCall(
+            channelToUse, safeGreeterSayHelloMethod, CallOptions.DEFAULT, message);
+
+    assertNotEquals(reply1, reply2);
+    assertEquals(0, cache.internalCache.size());
+    assertEquals(0, cache.removedKeys.size());
+  }
+
+  @Test
+  public void responseNoCache_caseInsensitive() throws Exception {
+    cacheControlDirectives.add("No-CaCHe");
+
+    HelloReply reply1 =
+        ClientCalls.blockingUnaryCall(
+            channelToUse, safeGreeterSayHelloMethod, CallOptions.DEFAULT, message);
+    HelloReply reply2 =
+        ClientCalls.blockingUnaryCall(
+            channelToUse, safeGreeterSayHelloMethod, CallOptions.DEFAULT, message);
+
+    assertNotEquals(reply1, reply2);
+    assertEquals(0, cache.internalCache.size());
+    assertEquals(0, cache.removedKeys.size());
+  }
+
+  @Test
+  public void combinedResponseCacheControlDirectives_parsesWithoutError() throws Exception {
+    cacheControlDirectives.add("max-age=1,no-store , no-cache");
+
+    HelloReply reply1 =
+        ClientCalls.blockingUnaryCall(
+            channelToUse, safeGreeterSayHelloMethod, CallOptions.DEFAULT, message);
+    HelloReply reply2 =
+        ClientCalls.blockingUnaryCall(
+            channelToUse, safeGreeterSayHelloMethod, CallOptions.DEFAULT, message);
+
+    assertNotEquals(reply1, reply2);
+    assertEquals(0, cache.internalCache.size());
+    assertEquals(0, cache.removedKeys.size());
+  }
+
+  @Test
+  public void separateResponseCacheControlDirectives_parsesWithoutError() throws Exception {
+    cacheControlDirectives.add("max-age=1");
+    cacheControlDirectives.add("no-store , no-cache");
+
+    HelloReply reply1 =
+        ClientCalls.blockingUnaryCall(
+            channelToUse, safeGreeterSayHelloMethod, CallOptions.DEFAULT, message);
+    HelloReply reply2 =
+        ClientCalls.blockingUnaryCall(
+            channelToUse, safeGreeterSayHelloMethod, CallOptions.DEFAULT, message);
+
+    assertNotEquals(reply1, reply2);
+    assertEquals(0, cache.internalCache.size());
+    assertEquals(0, cache.removedKeys.size());
+  }
+
+  @Test
+  public void afterResponseMaxAge_cacheEntryInvalidated() throws Exception {
+    cacheControlDirectives.add("max-age=1");
+
+    HelloReply reply1 =
+        ClientCalls.blockingUnaryCall(
+            channelToUse, safeGreeterSayHelloMethod, CallOptions.DEFAULT, message);
+    HelloReply reply2 =
+        ClientCalls.blockingUnaryCall(
+            channelToUse, safeGreeterSayHelloMethod, CallOptions.DEFAULT, message);
+    assertEquals(reply1, reply2);
+
+    // Wait for cache entry to expire
+    Thread.sleep(1001);
```

Comment:
needs to be done in a loop, it could sleep for less time.

Reply:
Also, since the clock timer is not super accurate, you probably need to sleep for a bit longer.  For example, the windows currentTimeMillis is only accurate to about 16ms.

Reply:
https://github.com/grpc/grpc-java/blob/f9f603570d20d4507f4d90a3dc53915ac7db4d33/testing/src/main/java/io/grpc/internal/testing/TestUtils.java#L171

Reply:
Done

Reply:
Not done?

Reply:
Hah, oops - I did manage to change one of these, but there were even more of them to be rooted out. Now it's really done.